python_modules/dagster/dagster_tests/asset_defs_tests/asset_graph_view_tests/test_bfs.py

```diff
@@ -0,0 +1,253 @@
+        [graph_view.get_full_subset(key=asset1.key), graph_view.get_full_subset(key=asset2.key)]
+    )
+    assert len(failed) == 1
+    assert failed[0][0] == AssetGraphSubset.from_entity_subsets(
+        [graph_view.get_full_subset(key=asset3.key)]
+    )
+    assert failed[0][1] == "asset3 not allowed"
+
+
+def test_bfs_filter_multi_asset():
+    """Test BFS filter with multi-asset."""
+
+    @asset
+    def a():
+        return 1
+
+    @asset
+    def b():
+        return 2
+
+    @multi_asset(
+        specs=[AssetSpec("c", deps=["a"]), AssetSpec("d", deps=["b"])]
+    )  # d is level 1, e is level 3, gets assigned 3
+    def my_multi_asset():
+        pass
+
+    @asset
+    def e(d):
+        pass
+
+    graph_view = AssetGraphView.for_test(Definitions(assets=[a, b, my_multi_asset, e]))
+    initial_subset = _get_subset_with_keys(graph_view, [a.key])
+
+    def condition_fn(subset, _visited):
+        return AssetGraphViewBfsFilterConditionResult(subset, [])
+
+    result_without_full_execution_set, _failed = bfs_filter_asset_graph_view(
+        graph_view, condition_fn, initial_subset, include_full_execution_set=False
+    )
+
+    assert result_without_full_execution_set == _get_subset_with_keys(
+        graph_view, [a.key, AssetKey(["c"])]
+    )
+
+    result_with_full_execution_set, _failed = bfs_filter_asset_graph_view(
+        graph_view, condition_fn, initial_subset, include_full_execution_set=True
+    )
+
+    assert result_with_full_execution_set == _get_subset_with_keys(
+        graph_view, [a.key, AssetKey(["c"]), AssetKey(["d"]), e.key]
+    )
+
+
+def test_bfs_filter_diamond():
+    """Test BFS filter with a diamond-shaped graph to ensure bottom node is visited once."""
+
+    @asset
+    def top():
+        return 1
+
+    @asset
+    def left(top):
+        return top + 1
+
+    @asset
+    def right(top):
+        return top + 2
+
+    @asset
+    def bottom(left, right):
+        return left + right
+
+    graph_view = AssetGraphView.for_test(Definitions(assets=[top, left, right, bottom]))
+    initial_subset = _get_subset_with_keys(graph_view, [AssetKey("top")])
+
+    visit_count = {}
+
+    def condition_fn(subset, _visited):
+        for key in subset.asset_keys:
+            visit_count[key] = visit_count.get(key, 0) + 1
+        return AssetGraphViewBfsFilterConditionResult(subset, [])
+
+    result, failed = bfs_filter_asset_graph_view(
+        graph_view, condition_fn, initial_subset, include_full_execution_set=True
+    )
+
+    # Each node should be visited exactly once
+    assert visit_count == {
+        AssetKey("bottom"): 1,
+        AssetKey("left"): 1,
+        AssetKey("right"): 1,
+        AssetKey("top"): 1,
+    }
+    assert result == _get_subset_with_keys(
+        graph_view, [AssetKey("top"), AssetKey("left"), AssetKey("right"), AssetKey("bottom")]
+    )
+    assert failed == []
+
+
+from dagster import AssetIn, StaticPartitionsDefinition
+
+
+def test_bfs_filter_with_partitions():
```

Comment:
Would be good to test time partitioned assets too, especially around asserting the partition sort key stuff

Reply:
added new test case